[AArch64][SME2] Preserve ZT0 state around function calls

kmclaughlin-arm opened the pull request:
If a function has ZT0 state and calls a function which does not
preserve ZT0, the caller must save and restore ZT0 around the call.
If the caller shares ZT0 state and the callee is not shared ZA, we must
additionally call SMSTOP/SMSTART ZA around the call.

This patch adds new AArch64ISDNodes for spilling & filling ZT0.
Where requiresPreservingZT0 is true, ZT0 state will be preserved
across a call.

kmclaughlin-arm committed:
[SME2][Clang] Add tests with ZT0 state

kmclaughlin-arm committed:
[AArch64][SME2] Preserve ZT0 state around function calls

If a function has ZT0 state and calls a function which does not
preserve ZT0, the caller must save and restore ZT0 around the call.
If the caller shares ZT0 state and the callee is not shared ZA, we must
additionally call SMSTOP/SMSTART ZA around the call.

This patch adds new AArch64ISDNodes for spilling & filling of ZT0.
Where requiresPreservingZT0 is true, ZT0 state will be preserved
across a call.

kmclaughlin-arm committed:
- Added requiresZAToggle() to AArch64SMEAttributes
- Added a test for an aarch64_in_zt0 caller -> aarch64_new_zt0 callee

kmclaughlin-arm committed:
- Run clang-format

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8074,14 +8105,20 @@ AArch64TargetLowering::LowerCall(CallLoweringInfo &CLI,
                                  PStateSM, false);
   }
 
+  if ((RequiresLazySave && !CalleeAttrs.preservesZA()) || ToggleZA)
```
Sorry for not thinking ahead on this one, but seeing this I just wondered if it would be easier to have two interfaces:
* `CallerAttrs.requiresDisablingZABeforeCall(CalleeAttrs)`
* `CallerAttrs.requiresEnablingZAAfterCall(CalleeAttrs)`

instead of `requiresZAToggle`?

sdesmalen-arm commented on llvm/lib/Target/AArch64/Utils/AArch64SMEAttributes.h:
```diff
@@ -119,6 +119,12 @@ class SMEAttrs {
            State == StateValue::InOut || State == StateValue::Preserved;
   }
   bool hasZT0State() const { return isNewZT0() || sharesZT0(); }
+  bool requiresPreservingZT0(const SMEAttrs &Callee) const {
```
Can we have unit tests for these in the file that tests this class?

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -7664,6 +7666,35 @@ AArch64TargetLowering::LowerCall(CallLoweringInfo &CLI,
 
+  SDValue ZTFrameIdx;
+  MachineFrameInfo &MFI = MF.getFrameInfo();
+  bool PreserveZT0 = CallerAttrs.requiresPreservingZT0(CalleeAttrs);
```
nit:
```suggestion
  bool ShouldPreserveZT0 = CallerAttrs.requiresPreservingZT0(CalleeAttrs);
```

To avoid confusion with `preserves("zt0")`

kmclaughlin-arm committed:
- Split requiresPreservingZT0 into requiresDisablingZABeforeCall/requiresEnablingZAAfterCall
- Renamed PreserveZT0 to ShouldPreserveZT0
- Added unittests for requiresPreservingZT0, requiresDisablingZABeforeCall requiresEnablingZAAfterCall

sdesmalen-arm commented on llvm/unittests/Target/AArch64/SMEAttributesTest.cpp:
```diff
@@ -275,4 +278,88 @@ TEST(SMEAttributes, Transitions) {
+  SA ZT0_New = SA(SA::encodeZT0State(SA::StateValue::New));
+
+  // ZT0 New -> Normal
+  ASSERT_TRUE(ZT0_New.requiresPreservingZT0(SA(SA::Normal)));
+  ASSERT_TRUE(ZT0_New.requiresDisablingZABeforeCall(SA(SA::Normal)));
+  ASSERT_TRUE(ZT0_New.requiresEnablingZAAfterCall(SA(SA::Normal)));
+
+  // ZT0 New -> ZT0 New
+  ASSERT_TRUE(ZT0_New.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_New.requiresDisablingZABeforeCall(ZT0_New));
+  ASSERT_TRUE(ZT0_New.requiresEnablingZAAfterCall(ZT0_New));
```
Normal and ZT0_New have the same interface, so there's no value in testing them both.

sdesmalen-arm commented on llvm/unittests/Target/AArch64/SMEAttributesTest.cpp:
```diff
@@ -275,4 +278,88 @@ TEST(SMEAttributes, Transitions) {
+  ASSERT_TRUE(ZT0_New.requiresEnablingZAAfterCall(ZT0_New));
+
+  // ZT0 New -> ZT0 Shared
+  ASSERT_FALSE(ZT0_New.requiresPreservingZT0(ZT0_In));
```
I don't see much value in testing all these different variations of "sharing ZT0" (ZT0_In/InOut/Out/Preserves) because `Callee.hasPrivateZAInterface()` and `Callee.sharesZT0()` will return `false` for all of them (which is also tested elsewhere), so they're all testing the same code-path + data for these functions.

sdesmalen-arm commented on llvm/unittests/Target/AArch64/SMEAttributesTest.cpp:
```diff
@@ -275,4 +278,88 @@ TEST(SMEAttributes, Transitions) {
+  ASSERT_TRUE(ZT0_Out.requiresDisablingZABeforeCall(SA(SA::Normal)));
+  ASSERT_TRUE(ZT0_Out.requiresEnablingZAAfterCall(SA(SA::Normal)));
+
+  ASSERT_TRUE(ZT0_Preserved.requiresPreservingZT0(SA(SA::Normal)));
+  ASSERT_TRUE(ZT0_Preserved.requiresDisablingZABeforeCall(SA(SA::Normal)));
+  ASSERT_TRUE(ZT0_Preserved.requiresEnablingZAAfterCall(SA(SA::Normal)));
+
+  // ZT0 Shared -> ZT0 Shared
+  ASSERT_FALSE(ZT0_In.requiresPreservingZT0(ZT0_In));
+  ASSERT_FALSE(ZT0_In.requiresDisablingZABeforeCall(ZT0_In));
+  ASSERT_FALSE(ZT0_In.requiresEnablingZAAfterCall(ZT0_In));
+
+  ASSERT_FALSE(ZT0_InOut.requiresPreservingZT0(ZT0_In));
+  ASSERT_FALSE(ZT0_InOut.requiresDisablingZABeforeCall(ZT0_In));
+  ASSERT_FALSE(ZT0_InOut.requiresEnablingZAAfterCall(ZT0_In));
+
+  ASSERT_FALSE(ZT0_Out.requiresPreservingZT0(ZT0_In));
+  ASSERT_FALSE(ZT0_Out.requiresDisablingZABeforeCall(ZT0_In));
+  ASSERT_FALSE(ZT0_Out.requiresEnablingZAAfterCall(ZT0_In));
+
+  ASSERT_FALSE(ZT0_Preserved.requiresPreservingZT0(ZT0_In));
+  ASSERT_FALSE(ZT0_Preserved.requiresDisablingZABeforeCall(ZT0_In));
+  ASSERT_FALSE(ZT0_Preserved.requiresEnablingZAAfterCall(ZT0_In));
+
+  // ZT0 Shared -> ZT0 New
+  ASSERT_TRUE(ZT0_In.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_In.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_In.requiresEnablingZAAfterCall(ZT0_New));
+
+  ASSERT_TRUE(ZT0_InOut.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_InOut.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_InOut.requiresEnablingZAAfterCall(ZT0_New));
+
+  ASSERT_TRUE(ZT0_Out.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_Out.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_Out.requiresEnablingZAAfterCall(ZT0_New));
+
+  ASSERT_TRUE(ZT0_Preserved.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_Preserved.requiresPreservingZT0(ZT0_New));
+  ASSERT_TRUE(ZT0_Preserved.requiresEnablingZAAfterCall(ZT0_New));
```
Similar comment for these cases, they're not really testing anything that isn't already tested by the by other tests. `ZT0_In/Out/InOut/Preserves/New` all return `true` for `hasZT0State()`, so the codepath and data will be the same for these functions.

I think it's more valuable to add some tests for ZA (because of the lazy-save mechanism) and perhaps even some in combination with ZT0.

kmclaughlin-arm committed:
- Reduced number of tests added to SMEAttributesTest & added some with ZA state

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/sme-zt0-state.ll:
```diff
@@ -0,0 +1,165 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sme2 -start-after=simplifycfg -enable-tail-merge=false -verify-machineinstrs < %s | FileCheck %s
+
+; Callee with no ZT state
+declare void @no_state_callee();
+
+; Callees with ZT0 state
+declare void @zt0_shared_callee() "aarch64_in_zt0";
+declare void @zt0_new_callee() "aarch64_new_zt0";
+
+; Callees with ZA state
+
+declare void @za_shared_callee() "aarch64_pstate_za_shared";
+declare void @za_zt0_shared_callee() "aarch64_pstate_za_shared" "aarch64_in_zt0";
```
Sorry I didn't spot this earlier, but you can create a single `declare void @callee()` and then add the attributes on the callsite. That avoids the need for all thees declarations.

sdesmalen-arm reviewed (commented):
One small comment on the test, but otherwise I'm happy with the PR.

kmclaughlin-arm committed:
- Rebased on main
- Used a single callee() function in sme-zt0-state.ll and added attributes at the callsite

sdesmalen-arm reviewed (approved):
LGTM

kmclaughlin-arm committed:
Merge branch 'main' into sme2-shared-za-zt0
